src/v/config/property_constraints.cc

```diff
@@ -0,0 +1,462 @@
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "config/property_constraints.h"
+
+#include "cluster/types.h"
+#include "config/configuration.h"
+#include "model/fundamental.h"
+
+#include <chrono>
+#include <stdexcept>
+#include <string>
+
+namespace config {
+
+std::string_view to_string_view(constraint_type type) {
+    switch (type) {
+    case constraint_type::none:
+        return "none";
+    case constraint_type::_restrict:
+        return "restrict";
+    case constraint_type::clamp:
+        return "clamp";
+    }
+
+    return "invalid";
+}
+
+template<>
+std::optional<constraint_type>
+from_string_view<constraint_type>(std::string_view sv) {
+    return string_switch<std::optional<constraint_type>>(sv)
+      .match("none", constraint_type::none)
+      .match("restrict", constraint_type::_restrict)
+      .match("clamp", constraint_type::clamp)
+      .default_match(constraint_type::none);
+}
+
+// This is where all the type conversions happen
+std::shared_ptr<constraint_validator_t> constraint_validator_map(
+  ss::sstring& name,
+  std::optional<ss::sstring> min_str_opt,
+  std::optional<ss::sstring> max_str_opt,
+  constraint_enabled_t enabled) {
+    auto& cfg = config::shard_local_cfg();
+
+    if (name == "log_cleanup_policy") {
+        return std::make_shared<
+          constraint_validator_enabled<model::cleanup_policy_bitflags>>(
+          std::move(enabled),
+          cfg.log_cleanup_policy.bind(),
+          [](const cluster::topic_configuration& topic_cfg) {
+              return topic_cfg.properties.cleanup_policy_bitflags
+                       ? *topic_cfg.properties.cleanup_policy_bitflags
+                       : model::cleanup_policy_bitflags::none;
+          },
+          [](
+            cluster::topic_configuration& topic_cfg,
+            model::cleanup_policy_bitflags& topic_val) {
+              topic_cfg.properties.cleanup_policy_bitflags = topic_val;
+          });
+    } else if (name == "log_retention_ms") {
+        std::optional<std::chrono::milliseconds> min_opt{std::nullopt},
+          max_opt{std::nullopt};
+        if (min_str_opt) {
+            min_opt = std::chrono::milliseconds{std::stol(*min_str_opt)};
+        }
+        if (max_str_opt) {
+            max_opt = std::chrono::milliseconds{std::stol(*max_str_opt)};
+        }
+        auto def_val = *cfg.log_retention_ms.default_value();
+        return std::make_shared<
+          constraint_validator_range<std::chrono::milliseconds>>(
+          std::move(min_opt),
+          std::move(max_opt),
+          [def_val{std::move(def_val)}](
+            const cluster::topic_configuration& topic_cfg) {
+              return topic_cfg.properties.retention_duration
+                         .has_optional_value()
```

Comment:
issue: This only returns true if it holds a `std::optional<T>`.  It doesn't validate that the optional itself has a value, so the next line `.value()` may throw if `retention_duration` holds an empty `std::optional`

Reply:
According to tristate.h, `has_optional_value()` returns true when the tristate is not disabled and the optional holds a value.
https://github.com/redpanda-data/redpanda/blob/db4e7f33e41c66cea98fc4541b2caa628d59f25d/src/v/tristate.h#L50C1-L52C6

This is also how the output operator checks for value in https://github.com/redpanda-data/redpanda/blob/db4e7f33e41c66cea98fc4541b2caa628d59f25d/src/v/tristate.h#L106

So I am not sure I understand the issue here.